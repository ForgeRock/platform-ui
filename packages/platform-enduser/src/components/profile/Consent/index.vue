<!-- Copyright (c) 2020-2021 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <FrListGroup
    v-show="mappings.length"
    :title="$t('pages.profile.consent.title')"
    :subtitle="$t('pages.profile.consent.subtitle')">
    <FrListItem
      v-for="mapping in mappings"
      :key="mapping.name"
      :collapsible="mapping.consented"
      :panel-shown="false">
      <template
        v-slot:list-item-header
        class="overflow:hidden;">
        <FrFallbackImage
          :src="mapping.icon"
          class="mr-3"
          width="24"
          height="24"
          :alt="mapping.name"
          fallback="settings_applications" />
        <div
          class="media-body"
          style="width: 100%;">
          <div class="d-block">
            <h6 class="my-0">
              {{ mapping.displayName }}
            </h6>
            <small class="text-muted">
              {{ mapping.subTitle }}
            </small>
          </div>
        </div>

        <div class="d-flex align-self-right">
          <BButton
            variant="link"
            @click.stop.prevent="showModal(mapping.name)">
            {{ $t(`pages.profile.consent.${mapping.consented ? 'deny' : 'allow'}`) }}
          </BButton>
          <div
            class="caret ml-2 my-auto list-group-item-action"
            v-if="mapping.consented">
            <i class="material-icons-outlined font-weight-bolder md-24 mb-1 pr-0 caret-down align-self-center">
              keyboard_arrow_down
            </i>
            <i class="material-icons-outlined font-weight-bolder md-24 mb-1 pr-0 caret-up align-self-center">
              keyboard_arrow_up
            </i>
          </div>
        </div>

        <BModal
          :id="mapping.name"
          :ref="mapping.name"
          size="md"
          cancel-variant="outline-secondary">
          <template v-slot:modal-header>
            <div class="d-flex w-100 h-100">
              <h5 class="modal-title align-self-center text-center">
                {{ mapping.modalHeader }}
              </h5>
              <button
                type="button"
                aria-label="Close"
                class="close"
                @click.stop.prevent="hideModal(mapping.name)">
                <i class="material-icons-outlined font-weight-bolder md-24 mb-1">
                  close
                </i>
              </button>
            </div>
          </template>

          <BContainer>
            <p
              v-if="mapping.consented"
              v-html="$t('pages.profile.consent.confirmDeny', {mappingName: mapping.displayName})" />
            <FrAccessLevel
              v-else
              :fields="mapping.fields" />
          </BContainer>

          <template v-slot:modal-footer>
            <div class="float-right">
              <BButton
                variant="outline-secondary mr-2"
                @click.stop.prevent="hideModal(mapping.name)">
                {{ $t('common.cancel') }}
              </BButton>
              <BButton
                :variant="mapping.consented ? 'danger' : 'primary'"
                @click.stop.prevent="toggleConsentAndHideModal(mapping)">
                {{ $t(`pages.profile.consent.${mapping.consented ? 'deny' : 'allow'}`) }}
              </BButton>
            </div>
          </template>
        </BModal>
      </template>

      <template v-slot:list-item-collapse-body>
        <FrAccessLevel :fields="mapping.fields" />
      </template>
    </FrListItem>
  </FrListGroup>
</template>

<script>
import {
  cloneDeep,
  find,
  first,
  isUndefined,
} from 'lodash';
import dayjs from 'dayjs';
import advancedFormat from 'dayjs/plugin/advancedFormat';
import { mapState } from 'vuex';
import ListGroup from '@forgerock/platform-shared/src/components/ListGroup/';
import ListItem from '@forgerock/platform-shared/src/components/ListItem/';
import RestMixin from '@forgerock/platform-shared/src/mixins/RestMixin';
import encodeQueryString from '@forgerock/platform-shared/src/utils/encodeQueryString';
import AccessLevel from '@/components/profile/AccessLevel';
import FallbackImage from '@/components/utils/FallbackImage';

dayjs.extend(advancedFormat);

/**
 * @description Controls the display of a users currently consented mappings (where their data is sent).
 *
 * @fires POST consent?_action=getConsentMappings - Gets a list of available mappings for consent in conjunction with the consent portion of a users profile this is
 * used to display which mappings are consented to and can be consented out of.
 *
 */
export default {
  name: 'Consent',
  mixins: [
    RestMixin,
  ],
  components: {
    FrListGroup: ListGroup,
    FrListItem: ListItem,
    FrAccessLevel: AccessLevel,
    FrFallbackImage: FallbackImage,
  },
  props: {
    consentedMappings: {
      type: Array,
      default: () => [],
    },
  },
  data() {
    return {
      consentableMappings: [],
    };
  },
  computed: {
    ...mapState({
      managedResource: (state) => state.UserStore.managedResource,
    }),
    mappings() {
      return this.consentableMappings.map((mapping) => {
        const consentedMapping = find(this.consentedMappings, { mapping: mapping.name });
        const mappingCopy = cloneDeep(mapping);

        let modalHeaderPath = 'pages.profile.consent.';

        mappingCopy.showDetails = false;

        if (!isUndefined(consentedMapping)) {
          mappingCopy.consented = true;
          mappingCopy.consentDate = consentedMapping.consentDate;
          modalHeaderPath += 'denyConsentHeader';
          mappingCopy.subTitle = `${this.$t('pages.profile.consent.authorized')} ${dayjs(mapping.consentDate).format('MMMM Do YYYY')}`;
        } else {
          mappingCopy.consented = false;
          modalHeaderPath += 'allowConsentHeader';
          mappingCopy.subTitle = this.$t('pages.profile.consent.notAuthorized');
        }

        mappingCopy.modalHeader = this.$t(modalHeaderPath);
        return mappingCopy;
      });
    },
  },
  created() {
    this.loadConsent();
  },
  methods: {
    showModal(name) {
      first(this.$refs[name]).show();
    },
    loadConsent() {
      const urlParams = { queryFilter: `/source eq "${this.managedResource}"` };
      this.getRequestService()
        .get(`consent${encodeQueryString(urlParams)}`)
        .then(({ data }) => {
          this.consentableMappings = data.result;
        });
    },
    toggleConsentAndHideModal(mapping) {
      this.toggleConsent(mapping);
      this.hideModal(mapping.name);
    },
    hideModal(name) {
      first(this.$refs[name]).hide();
    },
    generatePatch(mapping) {
      const { consentDate, name } = mapping;
      const value = {
        consentDate,
        mapping: name,
      };

      let field = '/consentedMappings';
      let operation;

      if (!mapping.consented) {
        value.consentDate = new Date().toISOString();
        field += '/-';
        operation = 'add';
      } else {
        operation = 'remove';
      }

      return [{ field, operation, value }];
    },
    toggleConsent(mapping) {
      this.$emit('updateProfile', this.generatePatch(mapping), { onSuccess: this.loadConsent() });
    },
  },
};
</script>
