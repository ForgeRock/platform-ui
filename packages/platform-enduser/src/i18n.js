/**
 * Copyright (c) 2020-2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { createI18n } from 'vue-i18n';
import { merge } from 'lodash';
import dateTimeFormats from '@forgerock/platform-shared/src/dateTimeFormats.json';

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  // Can't use workspace symlink due to jest issues
  const sharedLocales = require.context('../../platform-shared/src/locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      let generatedTranslation = locales(key);

      // Check for shared local and merge in if it exists
      const sharedLocaleExists = sharedLocales.keys().includes(key);
      if (sharedLocaleExists) {
        generatedTranslation = merge(sharedLocales(key), generatedTranslation);
      }

      messages[locale] = generatedTranslation;
    }
  });

  return messages;
}

export default createI18n({
  messages: loadLocaleMessages(),
  silentFallbackWarn: true,
  silentTranslationWarn: true,
  dateTimeFormats,
});
