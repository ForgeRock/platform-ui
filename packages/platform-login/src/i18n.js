/**
 * Copyright (c) 2020-2021 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Vue from 'vue';
import VueI18n from 'vue-i18n';
import { merge } from 'lodash';
import dateTimeFormats from '@forgerock/platform-shared/src/dateTimeFormats.json';

Vue.use(VueI18n);

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  // Can't use workspace symlink due to jest issues
  const sharedLocales = require.context('../../platform-shared/src/locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      let generatedTranslation = locales(key);

      // Check for shared local and merge in if it exists
      const sharedLocaleExists = sharedLocales.keys().includes(key);
      if (sharedLocaleExists) {
        generatedTranslation = merge(sharedLocales(key), generatedTranslation);
      }

      messages[locale] = generatedTranslation;
    }
  });

  return messages;
}

Vue.mixin({
  methods: {
    translationExists(path) {
      return this.$t(path) !== path;
    },
  },
});

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages(),
  silentTranslationWarn: true,
  dateTimeFormats,
});
