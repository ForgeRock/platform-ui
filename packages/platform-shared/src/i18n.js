/**
 * Copyright 2019 ForgeRock AS. All Rights Reserved
 *
 * Use of this code requires a commercial software license with ForgeRock AS.
 * or with one of its affiliates. All use shall be exclusively subject
 * to such license between the licensee and ForgeRock AS.
 */
import Vue from 'vue';
import VueI18n from 'vue-i18n';
import dateTimeFormats from './dateTimeFormats.json';

Vue.use(VueI18n);

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      messages[locale] = locales(key);
    }
  });

  return messages;
}

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages(),
  silentTranslationWarn: true,
  dateTimeFormats,
});
