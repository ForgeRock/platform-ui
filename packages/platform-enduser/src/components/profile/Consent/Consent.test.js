/**
 * Copyright 2019 ForgeRock AS. All Rights Reserved
 *
 * Use of this code requires a commercial software license with ForgeRock AS.
 * or with one of its affiliates. All use shall be exclusively subject
 * to such license between the licensee and ForgeRock AS.
 */
import BootstrapVue from 'bootstrap-vue';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import {
  find,
  first,
} from 'lodash';
import Consent from '@/components/profile/Consent';
import i18n from '@/i18n';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

Consent.created = jest.fn();

describe('Profile Consent Component', () => {
  const date = new Date().toISOString();
  const consentableMappings = [
    { name: 'test' },
    { name: 'test2' },
  ];

  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Consent, {
      localVue,
      i18n,
      propsData: {
        consentedMappings: [{ mapping: 'test', consentDate: date }],
      },
    });
  });

  afterEach(() => {
    wrapper = undefined;
  });
  describe('Consent.vue', () => {
    it('Consent page loaded', () => {
      expect(wrapper.name()).toBe('Consent');
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('computed#mappings', () => {
    it('should return the mappings that are populated with user data', () => {
      wrapper.setData({ consentableMappings });

      const { mappings } = wrapper.vm;
      const test = find(mappings, { name: 'test' });
      const test2 = find(mappings, { name: 'test2' });

      expect(Array.isArray(mappings)).toBe(true);
      expect(mappings.length).toBe(2);
      expect(test).toHaveProperty('consented');
      expect(test.consented).toBe(true);
      expect(test).toHaveProperty('consentDate');
      expect(test.consentDate).toBe(date);
      expect(test2).toHaveProperty('consented');
      expect(test2.consented).toBe(false);
      expect(test2).not.toHaveProperty('consentDate');
    });
  });

  describe('#toggleConsentAndHideModal', () => {
    it('should call #toggleConsent and #hideModal', () => {
      const hideSpy = jest.spyOn(wrapper.vm, 'hideModal');
      const toggleSpy = jest.spyOn(wrapper.vm, 'toggleConsent');

      wrapper.vm.$refs = { test: [{ hide: jest.fn() }] };
      wrapper.vm.toggleConsentAndHideModal({ name: 'test' });

      expect(hideSpy).toBeCalled();
      expect(toggleSpy).toBeCalled();
    });
  });

  describe('#generatePatch', () => {
    it('should generate an "add" patch given an unconsented mapping', () => {
      const mapping = {
        name: 'test',
        consented: false,
      };

      const patch = wrapper.vm.generatePatch(mapping);
      const addOp = first(patch);

      expect(typeof patch).toBe('object');
      expect(typeof addOp).toBe('object');
      expect(addOp).toHaveProperty('field');
      expect(addOp.field).toBe('/consentedMappings/-');
      expect(addOp).toHaveProperty('operation');
      expect(addOp.operation).toBe('add');
      expect(typeof addOp).toBe('object');
      expect(addOp.value).toHaveProperty('mapping');
      expect(addOp.value.mapping).toBe('test');
      expect(typeof addOp.value).toBe('object');
    });

    it('should generate a "remove" patch given a consented mapping', () => {
      const mapping = {
        name: 'test',
        consented: true,
        consentDate: new Date().toISOString(),
      };

      const patch = wrapper.vm.generatePatch(mapping);
      const addOp = first(patch);

      expect(typeof patch).toBe('object');
      expect(typeof addOp).toBe('object');
      expect(addOp).toHaveProperty('field');
      expect(addOp.field).toBe('/consentedMappings');
      expect(addOp).toHaveProperty('operation');
      expect(addOp.operation).toBe('remove');
      expect(typeof addOp).toBe('object');
      expect(addOp.value).toHaveProperty('mapping');
      expect(addOp.value.mapping).toBe('test');
      expect(typeof addOp.value).toBe('object');
      expect(addOp.value.consentDate).toBe(mapping.consentDate);
    });
  });

  describe('#toggleConsent', () => {
    it('should emit "updateProfile"', () => {
      wrapper.setMethods({ generatePatch: jest.fn() });
      wrapper.vm.toggleConsent();

      expect(wrapper.emitted().updateProfile.length).toBe(1);
    });
  });
});
