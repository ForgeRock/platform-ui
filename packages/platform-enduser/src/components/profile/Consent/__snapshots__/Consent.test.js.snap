// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Consent Component Consent.vue Consent page loaded 1`] = `<frlistgroup-stub title="Personal Data Sharing" subtitle="Control how your data is shared with third parties." style="display: none;"></frlistgroup-stub>`;
